File: tinyoraclaw-service/tinyoraclaw_service/api/queue.py
from fastapi import APIRouter, HTTPException, Query, Request

from ..models.queue import (
    EnqueueMessageRequest,
    EnqueueResponseRequest,
    FailMessageRequest,
)

router = APIRouter()


def _get_queue_service(request: Request):
    svc = request.app.state.queue_service
    if not svc:
        raise HTTPException(status_code=503, detail="Queue service not available")
    return svc


# ── Messages (incoming queue) ─────────────────────────────────────────────


@router.post("/api/queue/enqueue")
async def enqueue_message(data: EnqueueMessageRequest, request: Request):
    svc = _get_queue_service(request)
    row_id = await svc.enqueue_message(data)
    return {"id": row_id, "status": "pending"}


@router.get("/api/queue/next/{agent_id}")
async def claim_next_message(agent_id: str, request: Request):
    svc = _get_queue_service(request)
    msg = await svc.claim_next_message(agent_id)
    if msg is None:
        return {"message": None}
    return {"message": msg}


@router.patch("/api/queue/{row_id}/complete")
async def complete_message(row_id: int, request: Request):
    svc = _get_queue_service(request)
    await svc.complete_message(row_id)
    return {"status": "completed"}


@router.patch("/api/queue/{row_id}/fail")
async def fail_message(row_id: int, body: FailMessageRequest, request: Request):
    svc = _get_queue_service(request)
    await svc.fail_message(row_id, body.error)
    return {"status": "failed"}


@router.get("/api/queue/status")
async def get_queue_status(request: Request):
    svc = _get_queue_service(request)
    return await svc.get_queue_status()


@router.get("/api/queue/dead")
async def get_dead_messages(request: Request):
    svc = _get_queue_service(request)
    messages = await svc.get_dead_messages()
    return {"messages": messages}


@router.post("/api/queue/dead/{row_id}/retry")
async def retry_dead_message(row_id: int, request: Request):
    svc = _get_queue_service(request)
    success = await svc.retry_dead_message(row_id)
    if not success:
        raise HTTPException(status_code=404, detail="Dead message not found")
    return {"status": "retrying"}


@router.delete("/api/queue/dead/{row_id}")
async def delete_dead_message(row_id: int, request: Request):
    svc = _get_queue_service(request)
    success = await svc.delete_dead_message(row_id)
    if not success:
        raise HTTPException(status_code=404, detail="Dead message not found")
    return {"status": "deleted"}


@router.post("/api/queue/recover-stale")
async def recover_stale_messages(
    request: Request, threshold_ms: int = Query(600000, ge=0)
):
    svc = _get_queue_service(request)
    recovered = await svc.recover_stale_messages(threshold_ms=threshold_ms)
    return {"recovered": recovered}


@router.delete("/api/queue/prune/responses")
async def prune_acked_responses(
    request: Request, older_than_ms: int = Query(86400000, ge=0)
):
    svc = _get_queue_service(request)
    pruned = await svc.prune_acked_responses(older_than_ms=older_than_ms)
    return {"pruned": pruned}


@router.delete("/api/queue/prune/messages")
async def prune_completed_messages(
    request: Request, older_than_ms: int = Query(86400000, ge=0)
):
    svc = _get_queue_service(request)
    pruned = await svc.prune_completed_messages(older_than_ms=older_than_ms)
    return {"pruned": pruned}


@router.get("/api/queue/pending-agents")
async def get_pending_agents(request: Request):
    svc = _get_queue_service(request)
    agents = await svc.get_pending_agents()
    return {"agents": agents}


@router.get("/api/queue/agents")
async def get_agent_queue_status(request: Request):
    svc = _get_queue_service(request)
    agents = await svc.get_agent_queue_status()
    return {"agents": agents}


@router.get("/api/queue/processing")
async def get_processing_messages(request: Request):
    svc = _get_queue_service(request)
    messages = await svc.get_processing_messages()
    return {"messages": messages}


# ── Responses (outgoing queue) ────────────────────────────────────────────


@router.post("/api/responses/enqueue")
async def enqueue_response(data: EnqueueResponseRequest, request: Request):
    svc = _get_queue_service(request)
    row_id = await svc.enqueue_response(data)
    return {"id": row_id, "status": "pending"}


@router.get("/api/responses/pending")
async def get_responses_for_channel(
    request: Request, channel: str = Query(..., description="Channel name"),
):
    svc = _get_queue_service(request)
    responses = await svc.get_responses_for_channel(channel)
    return {"responses": responses}


@router.get("/api/responses/recent")
async def get_recent_responses(
    request: Request, limit: int = Query(20, ge=1, le=500),
):
    svc = _get_queue_service(request)
    responses = await svc.get_recent_responses(limit)
    return {"responses": responses}


@router.post("/api/responses/{response_id}/ack")
async def ack_response(response_id: int, request: Request):
    svc = _get_queue_service(request)
    await svc.ack_response(response_id)
    return {"status": "acked"}

File: packages/server/src/routes/queue.ts
import { Hono } from 'hono';
import {
    log, emitEvent, genId,
    getQueueStatus, getAgentQueueStatus, getRecentResponses, getResponsesForChannel,
    ackResponse, enqueueResponse,
    getDeadMessages, retryDeadMessage, deleteDeadMessage,
    getProcessingMessages, failMessage, getActiveAgentIds, killAgentProcess, queueEvents,
} from '@tinyagi/core';

function parseMaybeJson<T>(value: unknown): T | undefined {
    if (value == null) return undefined;
    if (typeof value === 'string') return JSON.parse(value) as T;
    return value as T;
}

export function createQueueRoutes() {
    const app = new Hono();

    // GET /api/queue/status
    app.get('/api/queue/status', async (c) => {
        const status = await getQueueStatus();
        return c.json({
            incoming: status.pending,
            queued: status.queued,
            processing: status.processing,
            completed: status.completed,
            dead: status.dead,
            outgoing: status.responsesPending,
        });
    });

    // GET /api/responses
    app.get('/api/responses', async (c) => {
        const limit = parseInt(c.req.query('limit') || '20', 10);
        const rows = await getRecentResponses(limit);
        return c.json(rows.map((r: any) => ({
            id: r.id,
            channel: r.channel,
            sender: r.sender,
            senderId: r.sender_id,
            message: r.message,
            originalMessage: r.original_message,
            timestamp: r.created_at,
            messageId: r.message_id,
            agent: r.agent,
            files: parseMaybeJson<string[]>(r.files),
        })));
    });

    // GET /api/responses/pending?channel=whatsapp
    app.get('/api/responses/pending', async (c) => {
        const channel = c.req.query('channel');
        if (!channel) return c.json({ error: 'channel query param required' }, 400);

        const rows = await getResponsesForChannel(channel);
        return c.json(rows.map((r: any) => ({
            id: r.id,
            channel: r.channel,
            sender: r.sender,
            senderId: r.sender_id,
            message: r.message,
            originalMessage: r.original_message,
            messageId: r.message_id,
            agent: r.agent,
            files: parseMaybeJson<string[]>(r.files),
            metadata: parseMaybeJson<Record<string, unknown>>(r.metadata),
        })));
    });

    // POST /api/responses — enqueue a proactive outgoing message
    app.post('/api/responses', async (c) => {
        const body = await c.req.json();
        const { channel, sender, senderId, message, agent, files } = body as {
            channel?: string; sender?: string; senderId?: string;
            message?: string; agent?: string; files?: string[];
        };

        if (!channel || !sender || !message) {
            return c.json({ error: 'channel, sender, and message are required' }, 400);
        }

        const messageId = genId('proactive');
        await enqueueResponse({
            channel,
            sender,
            senderId,
            message,
            originalMessage: '',
            messageId,
            agent,
            files: files && files.length > 0 ? files : undefined,
        });

        log('INFO', `[API] Proactive response enqueued for ${channel}/${sender}`);
        emitEvent('message:done', { channel, sender, messageId });
        return c.json({ ok: true, messageId });
    });

    // POST /api/responses/:id/ack — acknowledge a response
    app.post('/api/responses/:id/ack', async (c) => {
        const id = parseInt(c.req.param('id'), 10);
        await ackResponse(id);
        return c.json({ ok: true });
    });

    // GET /api/queue/agents — per-agent queue depth
    app.get('/api/queue/agents', async (c) => {
        return c.json(await getAgentQueueStatus());
    });

    // GET /api/queue/dead
    app.get('/api/queue/dead', async (c) => {
        const dead = await getDeadMessages();
        return c.json(dead.map((m: any) => ({
            id: m.id,
            data: {
                channel: m.channel,
                sender: m.sender,
                senderId: m.sender_id,
                message: m.message,
                messageId: m.message_id,
                agent: m.agent,
            },
            failedReason: m.last_error,
            attemptsMade: m.retry_count,
            timestamp: m.created_at,
        })));
    });

    // POST /api/queue/dead/:id/retry
    app.post('/api/queue/dead/:id/retry', async (c) => {
        const id = parseInt(c.req.param('id'), 10);
        const ok = await retryDeadMessage(id);
        if (!ok) return c.json({ error: 'dead message not found' }, 404);
        log('INFO', `[API] Dead message ${id} retried`);
        return c.json({ ok: true });
    });

    // DELETE /api/queue/dead/:id
    app.delete('/api/queue/dead/:id', async (c) => {
        const id = parseInt(c.req.param('id'), 10);
        const ok = await deleteDeadMessage(id);
        if (!ok) return c.json({ error: 'dead message not found' }, 404);
        log('INFO', `[API] Dead message ${id} deleted`);
        return c.json({ ok: true });
    });

    // GET /api/queue/processing — list active processing messages + process status
    app.get('/api/queue/processing', async (c) => {
        const activeAgents = new Set(getActiveAgentIds());
        const messages = await getProcessingMessages();
        return c.json(messages.map((m: any) => {
            const agent = m.agent || 'default';
            return {
                id: m.id,
                messageId: m.message_id,
                channel: m.channel,
                sender: m.sender,
                message: m.message,
                agent,
                status: m.status as 'queued' | 'processing',
                processAlive: activeAgents.has(agent),
                startedAt: m.updated_at,
                duration: Date.now() - m.updated_at,
            };
        }));
    });

    // POST /api/queue/processing/:id/kill — kill agent process + fail the message
    app.post('/api/queue/processing/:id/kill', async (c) => {
        const id = parseInt(c.req.param('id'), 10);
        const messages = await getProcessingMessages();
        const msg = messages.find((m: any) => m.id === id);
        if (!msg) return c.json({ error: 'processing message not found' }, 404);

        const agent = msg.agent || 'default';
        const killed = killAgentProcess(agent);
        await failMessage(id, 'Manually terminated by user');

        // Signal main loop to clear the agent chain
        queueEvents.emit('agent:killed', { agentId: agent });

        log('INFO', `[API] Killed agent session for ${agent} (message ${id}), process killed: ${killed}`);
        return c.json({ ok: true, agent, processKilled: killed });
    });

    return app;
}
